controllers/trainingjob/trainingjob_controller.go

```diff
@@ -18,411 +18,347 @@ package trainingjob
-	// ADD FINALIZER IF NOT EXISTS
-	////////////////////////////////////////////////////////////////////////////////////////////////////
-
-	// examine DeletionTimestamp to determine if object is under deletion
-	if state.ObjectMeta.DeletionTimestamp.IsZero() {
-		// The object is not being deleted. Add finalizer if not present
-		if !ContainsString(state.ObjectMeta.Finalizers, SageMakerResourceFinalizerName) {
-			state.ObjectMeta.Finalizers = append(state.ObjectMeta.Finalizers, SageMakerResourceFinalizerName)
-
-			log.Info("Adding finalizer")
-
-			prevGeneration := state.ObjectMeta.GetGeneration()
-			if err = r.Update(ctx, &state); err != nil {
-				return RequeueIfError(err)
-			}
-			return RequeueImmediatelyUnlessGenerationChanged(prevGeneration, state.ObjectMeta.GetGeneration())
+	if ctx.TrainingJobDescription == nil {
+		if controllers.HasDeletionTimestamp(ctx.TrainingJob.ObjectMeta) {
+			return r.cleanupAndRemoveFinalizer(ctx)
 		}
-	}
 
-	////////////////////////////////////////////////////////////////////////////////////////////////////
-	// UPDATE ETCD TO MATCH SM API STATE
-	////////////////////////////////////////////////////////////////////////////////////////////////////
+		if err = r.createTrainingJob(ctx); err != nil {
+			return r.updateStatusAndReturnError(ctx, ReconcilingTrainingJobStatus, errors.Wrap(err, "Unable to create training job"))
+		}
 
-	if !r.etcdMatchesSmApi(state, describeResponse) {
-		if err = r.updateJobStatus(ctx, log, state, trainingjobv1.TrainingJobStatus{
-			SageMakerTrainingJobName: *state.Spec.TrainingJobName,
-			TrainingJobStatus:        string(trainingJobDescription.TrainingJobStatus),
-			SecondaryStatus:          string(trainingJobDescription.SecondaryStatus),
-			LastCheckTime:            Now(),
-			CloudWatchLogUrl:         cwLogUrl,
-			Additional:               GetOrDefault(trainingJobDescription.FailureReason, ""),
-		}); err != nil {
-			log.Info("Error updating ETCD to sync with SM API state")
-			return RequeueAfterInterval(r.PollInterval, err)
+		if ctx.TrainingJobDescription, err = ctx.SageMakerClient.DescribeTrainingJob(ctx, ctx.TrainingJobName); err != nil {
+			return r.updateStatusAndReturnError(ctx, ReconcilingTrainingJobStatus, errors.Wrap(err, "Unable to describe SageMaker training job"))
 		}
-		return RequeueAfterInterval(r.PollInterval, nil)
 	}
 
-	//////////////////////////////////////////////////////////////////////////////////////////////////////////
-	// BASE CASES
-	//////////////////////////////////////////////////////////////////////////////////////////////////////////
+	//TODO: Convert it to tinyurl or even better can we expose CW url via API server proxy UI?
+	ctx.CloudWatchLogURL = "https://" + *ctx.TrainingJob.Spec.Region + ".console.aws.amazon.com/cloudwatch/home?region=" +
+		*ctx.TrainingJob.Spec.Region + "#ctx.LogStream:group=/aws/sagemaker/TrainingJobs;prefix=" +
+		*ctx.TrainingJobDescription.TrainingJobName + ";streamFilter=typeLogStreamPrefix"
```

Comment:
How does this get set in the `ctx.TrainingJob.Status`? I think it should be set directly on the `ctx.TrainingJob.Status.CloudWatchLogUrl`.

The reason this happened is because we used to instantiate a new struct every time we updated the status, but now we update select fields.